In [19]:
from sklearn.ensemble import VotingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.ensemble import RandomForestClassifier
import sklearn.datasets as datasets
import numpy as np
import matplotlib.pyplot as plt
from sklearn import metrics

I will use the digits dataset because it is relatively small, so it will
be fast to fit and estimate a model. I will use three different classifiers on their own and compare the results to see how well they do against the others. Then I will use a Voting Classifier to see if combining the three into a hard or soft vote will work better. The three estimators are Linear Regression, Gaussian Naive Bayes and Random Forest Classifier

In [20]:
# The following section of loading and preparing the digits dataset is taken from 
# http://scikit-learn.org/stable/auto_examples/classification/plot_digits_classification.html
digits = datasets.load_digits()

# The data that we are interested in is made of 8x8 images of digits, let's
# have a look at the first 3 images, stored in the `images` attribute of the
# dataset.  If we were working from image files, we could load them using
# pylab.imread.  Note that each image must have the same size. For these
# images, we know which digit they represent: it is given in the 'target' of
# the dataset.
images_and_labels = list(zip(digits.images, digits.target))
for index, (image, label) in enumerate(images_and_labels[:4]):
    plt.subplot(2, 4, index + 1)
    plt.axis('off')
    plt.imshow(image, cmap=plt.cm.gray_r, interpolation='nearest')
    plt.title('Training: %i' % label)

# To apply a classifier on this data, we need to flatten the image, to
# turn the data in a (samples, feature) matrix:
n_samples = len(digits.images)
data = digits.images.reshape((n_samples, -1))

X_train = data[:n_samples / 2]
y_train = digits.target[:n_samples / 2]
X_test = data[n_samples / 2:]
y_test = digits.target[n_samples / 2:]

In [21]:
lrclass = LogisticRegression( solver = "lbfgs")
lrclass.fit(X_train,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='lbfgs', tol=0.0001,
          verbose=0, warm_start=False)

In [22]:
expected = digits.target[n_samples / 2:]
predicted_lr = lrclass.predict(data[n_samples / 2:])

In [23]:
gausClass = GaussianNB()
gausClass.fit(X_train,y_train)

GaussianNB()

In [24]:
predicted_gc = gausClass.predict(data[n_samples / 2:])

In [45]:
randForestClass = RandomForestClassifier(max_depth=5,n_estimators=100)
randForestClass.fit(X_train,y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=5, max_features='auto', max_leaf_nodes=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=100, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

In [46]:
predicted_rfc = randForestClass.predict(data[n_samples / 2:])

In [30]:
print("Classification report for classifier %s:\n%s\n"
      % (lrclass, metrics.classification_report(expected, predicted_lr)))
print("Confusion matrix:\n%s" % metrics.confusion_matrix(expected, predicted_lr))


Classification report for classifier LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='lbfgs', tol=0.0001,
          verbose=0, warm_start=False):
             precision    recall  f1-score   support

          0       0.97      0.97      0.97        88
          1       0.88      0.88      0.88        91
          2       0.99      0.97      0.98        86
          3       0.99      0.82      0.90        91
          4       0.99      0.95      0.97        92
          5       0.83      0.91      0.87        91
          6       0.95      0.99      0.97        91
          7       0.98      0.89      0.93        89
          8       0.89      0.90      0.89        88
          9       0.82      0.96      0.88        92

avg / total       0.93      0.92      0.92       899


Confusion matrix:
[[85  0  0  0  1  1  1  0  0  0]
 [ 0 80  0

In [31]:
print("Classification report for classifier %s:\n%s\n"
      % (gausClass, metrics.classification_report(expected, predicted_gc)))
print("Confusion matrix:\n%s" % metrics.confusion_matrix(expected, predicted_gc))


Classification report for classifier GaussianNB():
             precision    recall  f1-score   support

          0       0.98      0.95      0.97        88
          1       0.81      0.74      0.77        91
          2       0.87      0.84      0.85        86
          3       0.88      0.79      0.83        91
          4       1.00      0.73      0.84        92
          5       0.70      0.81      0.76        91
          6       0.96      0.99      0.97        91
          7       0.65      0.81      0.72        89
          8       0.61      0.76      0.68        88
          9       0.77      0.66      0.71        92

avg / total       0.82      0.81      0.81       899


Confusion matrix:
[[84  0  0  0  0  2  0  0  1  1]
 [ 0 67  2  0  0  0  0  2 13  7]
 [ 0  8 72  0  0  1  2  0  3  0]
 [ 0  2  2 72  0  2  0  2  9  2]
 [ 1  0  0  0 67  0  0 22  1  1]
 [ 0  2  0  4  0 74  1  3  2  5]
 [ 0  1  0  0  0  0 90  0  0  0]
 [ 0  0  2  0  0 12  0 72  2  1]
 [ 0  2  5  0  0  9  0  4 6

In [47]:
print("Classification report for classifier %s:\n%s\n"
      % (randForestClass, metrics.classification_report(expected, predicted_rfc)))
print("Confusion matrix:\n%s" % metrics.confusion_matrix(expected, predicted_rfc))


Classification report for classifier RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=5, max_features='auto', max_leaf_nodes=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=100, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False):
             precision    recall  f1-score   support

          0       0.95      0.99      0.97        88
          1       0.91      0.81      0.86        91
          2       0.96      0.87      0.91        86
          3       0.83      0.85      0.84        91
          4       0.94      0.92      0.93        92
          5       0.87      0.88      0.87        91
          6       0.98      0.99      0.98        91
          7       0.91      1.00      0.95        89
          8       0.93      0.84      0.88        88
          9       0.79      0.88      0.83        92

avg / total       0.91

Ok the reports are in and it shows that for the digits dataset the Linear Regression model has the greatest overall accuracy, but the other two were better in certain cases. Even though Gausian Naives Bayes had poorer overall if was best in 0,4 classifiers. Random Forests was best in 1,5,8,9. So with three good models lets let's hoping the Voting Classifier can be better combining all three.

The first selection will be to use a hard voting scheme. That is each of the three cast a vote and majority wins.

In [52]:
vclHardVote = VotingClassifier(estimators=[('lr', lrclass), ('rf', randForestClass), ('gnb', gausClass)], voting='hard')
vclHardVote.fit(X_train,y_train)
predicted_vclh = vclHardVote.predict(data[n_samples / 2:])
print("Classification report for classifier %s:\n%s\n"
      % (vclHardVote, metrics.classification_report(expected, predicted_vclh)))
print("Confusion matrix:\n%s" % metrics.confusion_matrix(expected, predicted_vclh))


Classification report for classifier VotingClassifier(estimators=[('lr', LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='lbfgs', tol=0.0001,
          verbose=0, warm_start=False)), ('rf', RandomFores...ob_score=False, random_state=None, verbose=0,
            warm_start=False)), ('gnb', GaussianNB())],
         voting='hard', weights=None):
             precision    recall  f1-score   support

          0       0.96      0.99      0.97        88
          1       0.87      0.84      0.85        91
          2       0.96      0.93      0.95        86
          3       0.93      0.86      0.89        91
          4       0.98      0.95      0.96        92
          5       0.86      0.92      0.89        91
          6       0.98      0.99      0.98        91
          7       0.93      0.96      0.94        89
          8       0.88  

Ok looks like Hard Vote brought the strongest one down with two weaker ones. Let's see how soft voting and works compared to hard voting.

In [67]:
vclSoftVote = VotingClassifier(estimators=[('lr', lrclass), ('rf', randForestClass), ('gnb', gausClass)], voting='soft')
vclSoftVote.fit(X_train,y_train)
predicted_vcls = vclSoftVote.predict(data[n_samples / 2:])
print("Classification report for classifier %s:\n%s\n"
      % (vclSoftVote, metrics.classification_report(expected, predicted_vcls)))
print("Confusion matrix:\n%s" % metrics.confusion_matrix(expected, predicted_vcls))


Classification report for classifier VotingClassifier(estimators=[('lr', LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='lbfgs', tol=0.0001,
          verbose=0, warm_start=False)), ('rf', RandomFores...ob_score=False, random_state=None, verbose=0,
            warm_start=False)), ('gnb', GaussianNB())],
         voting='soft', weights=None):
             precision    recall  f1-score   support

          0       0.98      0.99      0.98        88
          1       0.90      0.82      0.86        91
          2       0.92      0.97      0.94        86
          3       0.96      0.85      0.90        91
          4       1.00      0.92      0.96        92
          5       0.89      0.90      0.90        91
          6       0.98      0.99      0.98        91
          7       0.84      0.96      0.89        89
          8       0.80  

Ok So soft voting wasn't quite as good either.
Up next is weighted voting. best used if the classifiers perform better than others, so is the case here.

In [134]:
vclSoftVoteWeight = VotingClassifier(estimators=[('lr', lrclass), ('rf', randForestClass), ('gnb', gausClass)], voting='soft',weights=[5,8,1])
vclSoftVoteWeight.fit(X_train,y_train)
predicted_vclsw = vclSoftVoteWeight.predict(data[n_samples / 2:])
print("Classification report for classifier %s:\n%s\n"
      % (vclSoftVoteWeight, metrics.classification_report(expected, predicted_vclsw)))
print("Confusion matrix:\n%s" % metrics.confusion_matrix(expected, predicted_vclsw))


Classification report for classifier VotingClassifier(estimators=[('lr', LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='lbfgs', tol=0.0001,
          verbose=0, warm_start=False)), ('rf', RandomFores...ob_score=False, random_state=None, verbose=0,
            warm_start=False)), ('gnb', GaussianNB())],
         voting='soft', weights=[5, 8, 1]):
             precision    recall  f1-score   support

          0       0.96      0.99      0.97        88
          1       0.92      0.87      0.89        91
          2       1.00      0.98      0.99        86
          3       0.99      0.84      0.90        91
          4       0.99      0.93      0.96        92
          5       0.86      0.93      0.89        91
          6       0.96      1.00      0.98        91
          7       0.98      0.93      0.95        89
          8       0

Hurrah! With enough tweaking I could get the Voting Classifier to beat the best of the three classifiers. It turns out that weighting the Random Tree Classifier higher than the others brought the score up the highest even though the Linear Regression performs better than it head to head. Since the Gaussian Naive Bayes was doing poorly in comparison I did try leaving it out, but that made it worse. Even though it only gets a weight of 1 it still made a difference. Final Weighting to bring the score up above 93%
Random Forest Classifier 8
Linear Regression 5
Gaussian Naive Bayes 1